pkg/registry/common/querycache/nse_server.go

```diff
@@ -0,0 +1,95 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package querycache adds possible to cache Find queries
+package querycache
+
+import (
+	"context"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/registry"
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/sdk/pkg/registry/core/next"
+	"github.com/networkservicemesh/sdk/pkg/registry/core/streamchannel"
+	"github.com/networkservicemesh/sdk/pkg/registry/memory"
+)
+
+type queryCacheNSEClient struct {
+	cache memory.NetworkServiceEndpointSyncMap
+}
+
+func (q *queryCacheNSEClient) Register(ctx context.Context, in *registry.NetworkServiceEndpoint, opts ...grpc.CallOption) (*registry.NetworkServiceEndpoint, error) {
+	return next.NetworkServiceEndpointRegistryClient(ctx).Register(ctx, in, opts...)
+}
+
+func (q *queryCacheNSEClient) Find(ctx context.Context, in *registry.NetworkServiceEndpointQuery, opts ...grpc.CallOption) (registry.NetworkServiceEndpointRegistry_FindClient, error) {
+	if in.Watch {
+		return next.NetworkServiceEndpointRegistryClient(ctx).Find(ctx, in, opts...)
+	}
+	if nse, ok := q.cache.Load(in.String()); ok {
+		resultCh := make(chan *registry.NetworkServiceEndpoint, 1)
+		resultCh <- nse
+		close(resultCh)
+		return streamchannel.NewNetworkServiceEndpointFindClient(ctx, resultCh), nil
+	}
+	client, err := next.NetworkServiceEndpointRegistryClient(ctx).Find(ctx, in, opts...)
+	if err != nil {
+		return nil, err
+	}
+	nses := registry.ReadNetworkServiceEndpointList(client)
+	resultCh := make(chan *registry.NetworkServiceEndpoint, len(nses))
+	for _, nse := range nses {
+		nseQuery := &registry.NetworkServiceEndpointQuery{
+			NetworkServiceEndpoint: &registry.NetworkServiceEndpoint{
+				Name: nse.Name,
+			},
+		}
+		resultCh <- nse
+		key := nseQuery.String()
+		q.cache.Store(key, nse)
+		go func() {
+			defer q.cache.Delete(key)
+			nseQuery.Watch = true
+			stream, err := next.NetworkServiceEndpointRegistryClient(ctx).Find(ctx, nseQuery, opts...)
+			if err != nil {
+				q.cache.Delete(key)
+			}
+			for update, err := stream.Recv(); err == nil; update, err = stream.Recv() {
+				if update.Name != nseQuery.NetworkServiceEndpoint.Name {
+					continue
+				}
+				if update.ExpirationTime != nil && update.ExpirationTime.Seconds < 0 {
+					break
+				}
+				q.cache.Store(key, update)
+			}
+		}()
+	}
+	close(resultCh)
+	return streamchannel.NewNetworkServiceEndpointFindClient(ctx, resultCh), nil
+}
+
+func (q *queryCacheNSEClient) Unregister(ctx context.Context, in *registry.NetworkServiceEndpoint, opts ...grpc.CallOption) (*empty.Empty, error) {
+	return next.NetworkServiceEndpointRegistryClient(ctx).Unregister(ctx, in, opts...)
+}
+
+// NewClient creates new querycache registry.NetworkServiceEndpointRegistryClient that caches all resolved NSEs
+// All cached NSE is
+func NewClient() registry.NetworkServiceEndpointRegistryClient {
```

Comment:
Why is this file called nse_server.go when it contains a client?

Reply:
Oops! Will fix!

Reply:
thanks! Fixed.